pimd: fix memory leaks in parent process after fork()

jaredmauch opened the pull request:
Fix memory leaks that occur when the parent process exits after fork() in frr_daemonize(). The parent process was not cleaning up memory allocated before fork(), specifically:

- router structure allocated in pim_router_init()
- MLAG resources (mlag_fifo and mlag_stream) allocated in pim_mlag_init()
- zclient structures (pim_zclient and pim_zlookup) allocated in pim_zebra_init()

Add a new frr_parent_cleanup hook that is called in the parent process before it exits. Register pim_parent_cleanup() to free all PIM-specific resources in the parent process.

Also add NULL pointer checks in pim_mlag_terminate() to safely handle cases where MLAG was not initialized (e.g., pim6d).

Related PR: #18718

Jafaral commented on pimd/pimd.c:
```diff
@@ -124,7 +126,7 @@ void pim_router_terminate(void)
 
 /* Cleanup function for parent process after fork() */
-void pim_parent_cleanup(void)
+int pim_parent_cleanup(void)
```
I don't see a reason to change this if the only value we return is 0.

anlancs committed:
bgpd: update source address for bgp neighbor

The two bgp ipv6 neighbors: A and B, A is with "neighbor B bfd" and B is
with "neighbor A bfd strict".

Ater bgp neighbor is established, then delete and add the global address
(2001::75/64) on A. BFD of A will wrongly use "fe80::x" as source address forever:

```
2025/12/24 01:50:52 BFD: [SSYGJ-9ZAE0] zclient: add local address 2001::75/64 (VRF 6)
2025/12/24 01:50:52 BGP: [HFMHR-E3VMR] Rx Intf address add VRF vrf1 IF enp1s0 addr 2001::75/64
2025/12/24 01:50:52 BGP: [WJWH8-SHS78] bgp_peer_bfd_update_source: address [2001::75->3001::77] to [fe80::2e53:4aff:fe00:677->3001::77] <-
2025/12/24 01:50:52 BGP: [TT430-NZ800] zclient_bfd_command: multi hop is configured, not sending interface
```

The reason is currently A doesn't update BFD source until bgp neighbor is established.
But B uses BFD's strict mode, doesn't response A's OPEN message until BFD is UP,
It is dead lock for A and B.  So the bgp neighbor between A and B will not be
established and BFD is keeping in DOWN status because A is with "fe80::x" as source address.

So updating the source address not only when the bgp neighbor is established,
but also when the tcp connection is established.

Signed-off-by: anlan_cs <anlan_cs@126.com>

Jafaral commented:
@greptile review

Jafaral commented:
@jaredmauch , looks like something went wrong in the last update. I don't see any pim changes. there is a small change in bgp instread. Can you please recheck?
